README.md
# Ember-cli-deploy-ssh-tunnel

> An ember-cli-deploy plugin to open an ssh tunnel during your deploy

[![](https://ember-cli-deploy.github.io/ember-cli-deploy-version-badges/plugins/ember-cli-deploy-ssh-tunnel.svg)](http://ember-cli-deploy.github.io/ember-cli-deploy-version-badges/)

This plugins opens an ssh tunnel forwarding a port to the host you specify to one of your local ports.

A typical use case is to pair this plugin with the [ember-cli-deploy-redis](https://github.com/zapnito/ember-cli-deploy-redis) plugin to reach to redis servers that for security reasons are not listening to their public IP.

## What is an ember-cli-deploy plugin?

A plugin is an addon that can be executed as a part of the ember-cli-deploy pipeline. A plugin will implement one or more of the ember-cli-deploy's pipeline hooks.

For more information on what plugins are and how they work, please refer to the [Plugin Documentation][1].

## Quick Start

To get up and running quickly, do the following:

- Install this plugin

```bash
$ ember install ember-cli-deploy-ssh-tunnel
```

- For typical usage with `ember-cli-deploy-redis`, place the following configuration into `config/deploy.js`.

```javascript
ENV['ssh-tunnel'] = {
  username: 'yourname',
  host: 'yourserver',

  // A unique port on your local machine to forward to on the remote
  // server.  This will be set to a random port between 49151 and 65535
  // by default.
  // srcPort: <49151 - 65535>,

  // The port that the remote redis server listens on. 6379 is the default value.
  // dstPort: 6379
};

ENV.redis.host = 'localhost':

// ember-cli-redis will connect to redis locally on `ENV['ssh-tunnel'].srcPort`.
// Transmissions on that port will be forwarded over SSH and be received on
// the remote machine on `ENV['ssh-tunnel'].dstPort`.
// With `ember-cli-deploy-redis >= 0.1.1` `ENV.redis.port` will default
// to `ENV['ssh-tunnel'].srcPort` if available. On lower versions
// you must explicitly set it to match `ENV['ssh-tunnel'].srcPort.
//
// ENV.redis.port = 49151;
```

- Run the pipeline

```bash
$ ember deploy
```

## Installation
Run the following command in your terminal:

```bash
ember install ember-cli-deploy-ssh-tunnel
```

## ember-cli-deploy Hooks Implemented

For detailed information on what plugin hooks are and how they work, please refer to the [Plugin Documentation][1].

- `configure`
- `willUpload`
- `didUpload`

## Configuration Options

For detailed information on how configuration of plugins works, please refer to the [Plugin Documentation][1].

### username (`required`)

The user for the ssh connection.

*Default:* `undefined`

### host (`required`)

The server to connect to.

*Default:* `undefined`

### dstPort

The port to forward from the server.

*Default:* `6379`

### dstHost

The host to forward to on the destination server.

*Default:* `localhost`

### srcPort

The local port for the forwarding.

*Default:* a random port between `49151` and `65535`

### port

The ssh port on the destination server.

*Default:* `22`

### privateKeyPath

The local path to your ssh private key.

*Default:* null

### password

Authorization string for the ssh connection.

*Default:* null

### tunnelClient

The client used to create the ssh tunnel. This allows the user the ability to use their own client for uploading instead of the one provided by this plugin.

*Default:* the tunnel provided by `tunnel-ssh`

## Authorization

ember-cli-deploy-ssh-tunnel uses the [tunnel-ssh](https://github.com/Finanzchef24-GmbH/tunnel-ssh) module to provide the SSH tunnel. Two options exist to configure tunnel-ssh from ember-cli-deploy-ssh-tunnel: `privateKeyPath` and `password`. By default, we assume you have created a public and private key and added it to ssh-agent as described in the [default GitHub setup](https://help.github.com/articles/generating-ssh-keys/).

If no authentication information is delivered to tunnel-ssh, it will [default to using ssh-agent](https://github.com/Finanzchef24-GmbH/tunnel-ssh), so it will default to using the default id_rsa keys generated as described in the GitHub article. This includes password-protected SSH keys. If you would like to use a different SSH key, set the `privateKeyPath` option:

```js
ENV['ssh-tunnel'] = {
  username: 'yourname',
  host: 'yourserver',
  privateKeyPath: '~/.ssh/another_key_rsa'
};
```

If you just want to use a password to tunnel, you can specify that as an option (we recommend using environmental variables in an .env file):

```js
ENV['ssh-tunnel'] = {
  username: 'yourname',
  host: 'yourserver',
  password: process.env.SSH_PASSWORD
};
```

NOTE: at this time, this plugin does not support setting a path to `privateKeyPath` to a key that has been encrypted with a password.

## Using `ember-cli-deploy-ssh-tunnel` with "bastion" servers

`ember-cli-deploy-ssh-tunnel` is commonly used to open a tunnel on a "bastion" server in order to access services behind a firewall. A discussion of the problem is available on the README for the [`ember-cli-deploy-redis`](https://github.com/ember-cli-deploy/ember-cli-deploy-redis#what-if-my-redis-server-isnt-publicly-accessible) plugin.

### Example: AWS ElastiCache & EC2
One of the common use cases for `ember-cli-deploy-ssh-tunnel` is need to connect to ElastiCache that is accessible only from EC2 instance. To deploy your `index.html` to ElastiCache you need:
* username and host you can SSH into your EC2 instance (e.g. `deploy@my-ec2-instance.amazon.com`)
* host and port of your ElastiCache that is accessible from this EC2 instance (e.g. `my-elasticache-instance.amazon.com:6379`)

First, you should confirm that ssh tunneling actually works on your localhost, try:

```
$ ssh -f -N -L6379:my-elasticache-instance.amazon.com:6379 deploy@my-ec2-instance.amazon.com
$ redis-cli
```

You should be able to see Redis CLI connected to your ElastiCache instance.

If it works, set following configuration in `ssh-tunnel`:

1. Set `host` to your EC2 instance (e.g. `my-ec2-instance.amazon.com`)
2. Set `username` to your EC2 user (e.g. `deploy`)
3. Set `dstHost` to your ElastiCache instance (e.g. `my-elasticache-instance.amazon.com`)
4. Set `dstPort` to your ElastiCache port (e.g. `6379`)

In `redis` config leave `host` and `port` as default - thanks to ssh tunneling your `localhost:6379` will point to `my-elasticache-instance.amazon.com:6379` via `my-ec2-instance.amazon.com`.

## Running Tests

* yarn test

## Why `ember build` and `ember test` don't work

Since this is a node-only ember-cli addon, this package does not include many files and dependencies which are part of ember-cli's typical `ember build` and `ember test` processes.

## Thanks to:

@lukemelia and @achambers and the other folks from the ember-cli-deploy project.

@tim-evans for the original implementation in [ember-deploy-redis](https://github.com/LevelbossMike/ember-deploy-redis)

[1]: http://ember-cli.github.io/ember-cli-deploy/plugins "Plugin Documentation"


CHANGELOG.md
# Change Log


## v2.0.0 (2023-10-02)

#### :boom: Breaking Change
* [#12](https://github.com/ember-cli-deploy/ember-cli-deploy-ssh-tunnel/pull/12) [BREAKING] Update CI, dependencies & release process, require node >= 14 ([@lukemelia](https://github.com/lukemelia))

#### :rocket: Enhancement
* [#11](https://github.com/ember-cli-deploy/ember-cli-deploy-ssh-tunnel/pull/11) Update tunnel ssh ([@evoactivity](https://github.com/evoactivity))

#### Committers: 2
- Liam Potter ([@evoactivity](https://github.com/evoactivity))
- Luke Melia ([@lukemelia](https://github.com/lukemelia))

## [v1.0.0](https://github.com/ember-cli-deploy/ember-cli-deploy-ssh-tunnel/tree/v1.0.0) (2017-04-06)
[Full Changelog](https://github.com/ember-cli-deploy/ember-cli-deploy-ssh-tunnel/compare/v1.0.0-beta.0...v1.0.0)

**Merged pull requests:**

- \[DOC\] add ec2 & elasticache example usecase [\#7](https://github.com/ember-cli-deploy/ember-cli-deploy-ssh-tunnel/pull/7) ([jniechcial](https://github.com/jniechcial))

## [v1.0.0-beta.0](https://github.com/ember-cli-deploy/ember-cli-deploy-ssh-tunnel/tree/v1.0.0-beta.0) (2017-03-25)
[Full Changelog](https://github.com/ember-cli-deploy/ember-cli-deploy-ssh-tunnel/compare/v0.2.2...v1.0.0-beta.0)

**Merged pull requests:**

- Upgrade ember-cli and embrace being a node-only ember-cli addon [\#8](https://github.com/ember-cli-deploy/ember-cli-deploy-ssh-tunnel/pull/8) ([lukemelia](https://github.com/lukemelia))

## [v0.2.2](https://github.com/ember-cli-deploy/ember-cli-deploy-ssh-tunnel/tree/v0.2.2) (2016-02-06)
[Full Changelog](https://github.com/ember-cli-deploy/ember-cli-deploy-ssh-tunnel/compare/v0.2.1...v0.2.2)

**Merged pull requests:**

- Add more detail in the quickstart guide for setup with ember-cli-deploy-redis [\#5](https://github.com/ember-cli-deploy/ember-cli-deploy-ssh-tunnel/pull/5) ([g-cassie](https://github.com/g-cassie))

## [v0.2.1](https://github.com/ember-cli-deploy/ember-cli-deploy-ssh-tunnel/tree/v0.2.1) (2015-12-08)
[Full Changelog](https://github.com/ember-cli-deploy/ember-cli-deploy-ssh-tunnel/compare/v0.2.0...v0.2.1)

**Merged pull requests:**

- Defaults to ssh-agent instead of privateKeyPath [\#4](https://github.com/ember-cli-deploy/ember-cli-deploy-ssh-tunnel/pull/4) ([typeoneerror](https://github.com/typeoneerror))
- allow custom remote ssh port [\#3](https://github.com/ember-cli-deploy/ember-cli-deploy-ssh-tunnel/pull/3) ([bastianl](https://github.com/bastianl))

## [v0.2.0](https://github.com/ember-cli-deploy/ember-cli-deploy-ssh-tunnel/tree/v0.2.0) (2015-10-25)
[Full Changelog](https://github.com/ember-cli-deploy/ember-cli-deploy-ssh-tunnel/compare/v0.1.0...v0.2.0)

## [v0.1.0](https://github.com/ember-cli-deploy/ember-cli-deploy-ssh-tunnel/tree/v0.1.0) (2015-10-14)
**Merged pull requests:**

- Adds the ability to specify  for tunnel-ssh config [\#1](https://github.com/ember-cli-deploy/ember-cli-deploy-ssh-tunnel/pull/1) ([nikz](https://github.com/nikz))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

tests/runner.js
/*eslint-env node*/
'use strict';

var glob = require('glob');
var Mocha = require('mocha');

var mocha = new Mocha({
  reporter: 'spec'
});

var arg = process.argv[2];
var root = 'tests/';

function addFiles(mocha, files) {
  glob.sync(root + files).forEach(mocha.addFile.bind(mocha));
}

addFiles(mocha, '/**/*-test.js');

if (arg === 'all') {
  addFiles(mocha, '/**/*-test-slow.js');
}

mocha.run(function(failures) {
  process.on('exit', function() {
    // eslint-disable-next-line no-process-exit -- this sets the unix return code for the process
    process.exit(failures);
  });
});


tests/unit/index-test.js
'use strict';

var assert = require('../helpers/assert');

describe('ssh-tunnel plugin', function() {
  var subject, mockUi;

  beforeEach(function() {
    subject = require('../../index');
    mockUi = {
      messages: [],
      write: function() { },
      writeLine: function(message) {
        this.messages.push(message);
      },
      verbose: true
    };
  });

  it('has a name', function() {
    var result = subject.createDeployPlugin({
      name: 'test-plugin'
    });

    assert.equal(result.name, 'test-plugin');
  });

  it('implements the correct hooks', function() {
    var result = subject.createDeployPlugin({
      name: 'test-plugin'
    });

    assert.equal(typeof result.defaultConfig, 'object');
    assert.equal(typeof result.requiredConfig, 'object');
    assert.equal(typeof result.setup, 'function');
    assert.equal(typeof result.teardown, 'function');
  });

  describe('configure hook', function() {

    it('resolves if config is ok', function() {
      var plugin = subject.createDeployPlugin({
        name: 'ssh-tunnel'
      });

      var context = {
        ui: mockUi,
        config: {
          'ssh-tunnel': {
            host: 'example.com',
            username: 'ghedamat'
          }
        }
      };

      plugin.beforeHook(context);
      plugin.configure(context);
      assert.ok(true); // it didn't throw
    });

    describe('without providing config', function() {
      var plugin, context, config;

      beforeEach(function() {
        plugin = subject.createDeployPlugin({
          name: 'ssh-tunnel'
        });
      });

      it('raises about missing required config', async function() {
        config = {};
        context = {
          ui: mockUi,
          config: config
        };
        plugin.beforeHook(context);
        await assert.throws(function(){
          plugin.configure(context);
        }, /Missing required/);
        var messages = mockUi.messages.reduce(function(previous, current) {
          if (/- Missing required config:\s.*/.test(current)) {
            previous.push(current);
          }
          return previous;
        }, []);
        assert.equal(messages.length, 1);
      });

      it('warns about missing optional config', function() {
        context = {
          ui: mockUi,
          config: {
            'ssh-tunnel': {
              host: 'example.com',
              username: 'ghedamat'
            }
          }
        };
        plugin.beforeHook(context);
        plugin.configure(context);
        var messages = mockUi.messages.reduce(function(previous, current) {
          if (/- Missing config:\s.*, using default:\s/.test(current)) {
            previous.push(current);
          }

          return previous;
        }, []);
        assert.equal(messages.length, 5);
      });

      it('adds default config to the config object', function() {
        config = {
          'ssh-tunnel': {
            host: 'example.com',
            username: 'ghedamat'
          }
        };
        context = {
          ui: mockUi,
          config: config
        };
        plugin.beforeHook(context);
        plugin.configure(context);
        assert.isDefined(config['ssh-tunnel'].dstPort);
        assert.isDefined(config['ssh-tunnel'].dstHost);
        assert.isDefined(config['ssh-tunnel'].srcPort);
        assert.isDefined(config['ssh-tunnel'].tunnelClient);
      });
    });

    describe('with custom authentication provided', function () {
      var plugin, context, config;

      beforeEach(function() {
        plugin = subject.createDeployPlugin({
          name: 'ssh-tunnel'
        });
      });

      it('uses a password', function() {
        config = {
          'ssh-tunnel': {
            host: 'example.com',
            username: 'example',
            password: 'secret'
          }
        };
        context = {
          ui: mockUi,
          config: config
        };
        plugin.beforeHook(context);
        plugin.configure(context);
        assert.isDefined(config['ssh-tunnel'].password);
        assert.isUndefined(config['ssh-tunnel'].privateKeyPath);
      });

      it('uses a key path', function() {
        config = {
          'ssh-tunnel': {
            host: 'example.com',
            username: 'example',
            privateKeyPath: '~/.ssh/id_rsa'
          }
        };
        context = {
          ui: mockUi,
          config: config
        };
        plugin.beforeHook(context);
        plugin.configure(context);
        assert.isUndefined(config['ssh-tunnel'].password);
        assert.isDefined(config['ssh-tunnel'].privateKeyPath);
        assert.equal(config['ssh-tunnel'].privateKeyPath, '~/.ssh/id_rsa');
      });
    });
  });

  describe('setup hook', function() {
    var plugin, context;

    beforeEach(function() {
      plugin = subject.createDeployPlugin({
        name: 'ssh-tunnel'
      });

      context = {
        ui: mockUi,
        config: {
          'ssh-tunnel': {
            host: 'example.com',
            username: 'ghedamat',
            srcPort: 50000,
            dstPort: 6379,
            tunnelClient: function() {
              return function(config, cbk) {
                process.nextTick(cbk, false, true);
                return {
                  close: function() { }
                };
              };
            }
          }
        },
      };
      plugin.beforeHook(context);
    });

    it('returns the updated context if tunnel is successful', function(done) {
      assert.isFulfilled(plugin.setup(context)).then(function(result) {
        assert.isDefined(result.tunnel.handler);
        assert.isDefined(result.tunnel.srcPort);
        done();
      }).catch(function(reason) {
        done(reason.actual.stack);
      });
    });
  });

  describe('teardown hook', function() {
    var plugin, context;

    beforeEach(function() {
      plugin = subject.createDeployPlugin({
        name: 'ssh-tunnel'
      });

      context = {
        ui: mockUi,
        config: {
          'ssh-tunnel': {
          }
        }
      };
      plugin.beforeHook(context);
    });

    it('closes the tunnel', function(done) {
      context.tunnel = {
        handler: {
          close: function() {
            assert.ok(true);
            done();
          }
        }
      };

      plugin.teardown(context);
    });
  });
});


tests/unit/index-nodetest.js
/*jshint globalstrict: true*/
'use strict';

var assert  = require('ember-cli/tests/helpers/assert');

describe('ssh-tunnel plugin', function() {
  var subject, mockUi;

  beforeEach(function() {
    subject = require('../../index');
    mockUi = {
      messages: [],
      write: function() { },
      writeLine: function(message) {
        this.messages.push(message);
      },
      verbose: true
    };
  });

  it('has a name', function() {
    var result = subject.createDeployPlugin({
      name: 'test-plugin'
    });

    assert.equal(result.name, 'test-plugin');
  });

  it('implements the correct hooks', function() {
    var result = subject.createDeployPlugin({
      name: 'test-plugin'
    });

    assert.equal(typeof result.defaultConfig, 'object');
    assert.equal(typeof result.requiredConfig, 'object');
    assert.equal(typeof result.setup, 'function');
    assert.equal(typeof result.teardown, 'function');
  });

  describe('configure hook', function() {

    it('resolves if config is ok', function() {
      var plugin = subject.createDeployPlugin({
        name: 'ssh-tunnel'
      });

      var context = {
        ui: mockUi,
        config: {
          'ssh-tunnel': {
            host: 'example.com',
            username: 'ghedamat'
          }
        }
      };

      plugin.beforeHook(context);
      plugin.configure(context);
      assert.ok(true); // it didn't throw
    });

    describe('without providing config', function() {
      var plugin, context, config;

      beforeEach(function() {
        plugin = subject.createDeployPlugin({
          name: 'ssh-tunnel'
        });
      });

      it('raises about missing required config', function() {
        config = {};
        context = {
          ui: mockUi,
          config: config
        };
        plugin.beforeHook(context);
        assert.throws(function(){
          plugin.configure(context);
        });
        var messages = mockUi.messages.reduce(function(previous, current) {
          if (/- Missing required config:\s.*/.test(current)) {
            previous.push(current);
          }
          return previous;
        }, []);
        assert.equal(messages.length, 1);
      });

      it('warns about missing optional config', function() {
        context = {
          ui: mockUi,
          config: {
            'ssh-tunnel': {
              host: 'example.com',
              username: 'ghedamat'
            }
          }
        };
        plugin.beforeHook(context);
        plugin.configure(context);
        var messages = mockUi.messages.reduce(function(previous, current) {
          if (/- Missing config:\s.*, using default:\s/.test(current)) {
            previous.push(current);
          }

          return previous;
        }, []);
        assert.equal(messages.length, 5);
      });

      it('adds default config to the config object', function() {
        config = {
          'ssh-tunnel': {
            host: 'example.com',
            username: 'ghedamat'
          }
        };
        context = {
          ui: mockUi,
          config: config
        };
        plugin.beforeHook(context);
        plugin.configure(context);
        assert.isDefined(config['ssh-tunnel'].dstPort);
        assert.isDefined(config['ssh-tunnel'].dstHost);
        assert.isDefined(config['ssh-tunnel'].srcPort);
        assert.isDefined(config['ssh-tunnel'].tunnelClient);
      });
    });

    describe('with custom authentication provided', function () {
      var plugin, context, config;

      beforeEach(function() {
        plugin = subject.createDeployPlugin({
          name: 'ssh-tunnel'
        });
      });

      it('uses a password', function() {
        config = {
          'ssh-tunnel': {
            host: 'example.com',
            username: 'example',
            password: 'secret'
          }
        };
        context = {
          ui: mockUi,
          config: config
        };
        plugin.beforeHook(context);
        plugin.configure(context);
        assert.isDefined(config['ssh-tunnel'].password);
        assert.isUndefined(config['ssh-tunnel'].privateKeyPath);
      });

      it('uses a key path', function() {
        config = {
          'ssh-tunnel': {
            host: 'example.com',
            username: 'example',
            privateKeyPath: '~/.ssh/id_rsa'
          }
        };
        context = {
          ui: mockUi,
          config: config
        };
        plugin.beforeHook(context);
        plugin.configure(context);
        assert.isUndefined(config['ssh-tunnel'].password);
        assert.isDefined(config['ssh-tunnel'].privateKeyPath);
        assert.equal(config['ssh-tunnel'].privateKeyPath, '~/.ssh/id_rsa');
      });
    });
  });

  describe('setup hook', function() {
    var plugin, context;

    beforeEach(function() {
      plugin = subject.createDeployPlugin({
        name: 'ssh-tunnel'
      });

      context = {
        ui: mockUi,
        config: {
          'ssh-tunnel': {
            host: 'example.com',
            username: 'ghedamat',
            srcPort: 50000,
            dstPort: 6379,
            tunnelClient: function() {
              return function(config, cbk) {
                process.nextTick(cbk, false, true);
                return {
                  close: function() { }
                };
              };
            }
          }
        },
      };
      plugin.beforeHook(context);
    });

    it('returns the updated context if tunnel is successful', function(done) {
      assert.isFulfilled(plugin.setup(context)).then(function(result) {
        assert.isDefined(result.tunnel.handler);
        assert.isDefined(result.tunnel.srcPort);
        done();
      }).catch(function(reason) {
        done(reason.actual.stack);
      });
    });
  });

  describe('teardown hook', function() {
    var plugin, context;

    beforeEach(function() {
      plugin = subject.createDeployPlugin({
        name: 'ssh-tunnel'
      });

      context = {
        ui: mockUi,
        config: {
          'ssh-tunnel': {
          }
        }
      };
      plugin.beforeHook(context);
    });

    it('closes the tunnel', function(done) {
      context.tunnel = {
        handler: {
          close: function() {
            assert.ok(true);
            done();
          }
        }
      };

      plugin.teardown(context);
    });
  });
});


.eslintrc.js
module.exports = {
  root: true,
  parser: '@babel/eslint-parser',
  parserOptions: {
    ecmaVersion: 2020,
    sourceType: 'module',
    requireConfigFile: false,
    babelOptions: {
      configFile: false,
      // your babel options
      presets: ["@babel/preset-env"],
    },
  },
  plugins: [
    'ember'
  ],
  extends: [
    'eslint:recommended',
    'plugin:ember/recommended'
  ],
  env: {
    browser: true
  },
  rules: {
  },
  overrides: [
    // node files
    {
      files: [
        '.eslintrc.js',
        '.template-lintrc.js',
        'ember-cli-build.js',
        'index.js',
        'testem.js',
        'blueprints/*/index.js',
        'config/**/*.js',
        'tests/**/*.js'
      ],
      parserOptions: {
        sourceType: 'script'
      },
      env: {
        browser: false,
        node: true
      },
      plugins: ['node'],
      extends: ['plugin:node/recommended'],
      rules: Object.assign({}, require('eslint-plugin-node').configs.recommended.rules, {
        // add your custom rules and overrides for node files here
        'ember/avoid-leaking-state-in-ember-objects': 'off'
      })
    }
  ]
};


index.js
'use strict';

const RSVP = require('rsvp');
const fs = require('fs');
const { createTunnel } = require('tunnel-ssh');
const untildify = require('untildify');

const DeployPluginBase = require('ember-cli-deploy-plugin');

const MAX_PORT_NUMBER = 65535;
const MIN_PORT_NUMBER = 49151;

module.exports = {
  name: 'ember-cli-deploy-ssh-tunnel',

  createDeployPlugin: function (options) {
    const DeployPlugin = DeployPluginBase.extend({
      name: options.name,
      defaultConfig: {
        dstPort: 6379,
        port: 22,
        dstHost: 'localhost',
        srcPort: function () {
          var range = MAX_PORT_NUMBER - MIN_PORT_NUMBER + 1;
          return Math.floor(Math.random() * range) + MIN_PORT_NUMBER;
        },
        tunnelClient: function (context) {
          // if you want to provide your own ssh client to be used instead of one from this plugin,
          // must follow this signature 
          // createTunnel(
          //   tunnelOptions: TunnelOptions,
          //   serverOptions: ServerOptions,
          //   sshOptions: SshOptions,
          //   forwardOptions: ForwardOptions
          // ): Promise<[Server, Client]>;
          // https://github.com/agebrock/tunnel-ssh/blob/master/types/index.d.ts
          return context.tunnelClient || createTunnel;
        }
      },

      requiredConfig: ['host', 'username'],

      setup: function (/* context */) {
        const srcPort = this.readConfig('srcPort');

        if (srcPort > MAX_PORT_NUMBER || srcPort < MIN_PORT_NUMBER) {
          throw (
            'Port ' +
            srcPort +
            ' is not available to open a SSH connection on.\n' +
            'Please choose a port between ' +
            MIN_PORT_NUMBER +
            ' and ' +
            MAX_PORT_NUMBER +
            '.'
          );
        }

        const tunnel = this.readConfig('tunnelClient');

        let privateKey = this.readConfig('privateKey');

        if (this.readConfig('privateKeyPath')) {
          privateKey = fs.readFileSync(untildify(this.readConfig('privateKeyPath')));
        }

        const tunnelOptions = {
          autoClose: true
        };

        const serverOptions = {
          port: srcPort
        };

        const sshOptions = {
          host: this.readConfig('host'),
          port: this.readConfig('port'),
          username: this.readConfig('username'),
          password: this.readConfig('password'),
          privateKey,
          passphrase: this.readConfig('passphrase')
        };

        const forwardOptions = {
          srcAddr: 'localhost',
          srcPort: this.readConfig('srcPort'),
          dstAddr: this.readConfig('dstHost'),
          dstPort: this.readConfig('dstPort')
        };

        return new RSVP.Promise(function (resolve, reject) {
          tunnel(tunnelOptions, serverOptions, sshOptions, forwardOptions)
            .then(([server]) => {
              resolve({
                tunnel: {
                  handler: server,
                  srcPort: srcPort
                }
              });
            })
            .catch((error) => {
              reject(error);
            });
        });
      },

      teardown: function (context) {
        context.tunnel.handler.close();
      }
    });
    return new DeployPlugin();
  }
};
